### PR TITLE
`leanpkg build` without external dependencies

### PR DESCRIPTION
**Pull request by Kha**
This PR replaces `leanpkg`'s `make` dependency with an internal build system based on `Task` *if* the invocation is just `leanpkg build` or `leanpkg build -- lean-args...` and not something like `leanpkg build lib` that needs an external C compiler anyway. The directory structure is the identical, so mixing the two build systems should not lead to unnecessary rebuilds.

Moving from mtimes to hashing is TBD.

### CODE REVIEW

File: src/Leanpkg.lean [hunk 1]
```diff
@@ -88,16 +100,26 @@ def buildImports (imports : List String) (leanArgs : List String) : IO Unit := d
...
   IO.println cfg.leanPath
   IO.println cfg.leanSrcPath
 
 def build (makeArgs leanArgs : List String) : IO Unit := do
-  execMake makeArgs leanArgs (← configure).leanPath
+  let cfg ← configure
+  let root ← getRootPart
+  let buildCfg : Build.Config := { pkg := root, leanArgs, leanPath := cfg.leanPath, moreDeps := cfg.moreDeps }
+  if makeArgs != [] || (← FilePath.pathExists "Makefile") then
+    execMake leanArgs buildCfg
```
Comment (tydeu) [inline src/Leanpkg.lean:120]: I am pretty sure this should be `makeArgs` not `leanArgs`.

Comment (Kha) [reply src/Leanpkg.lean]: Definitely, thanks! I've extended the (single) test case to test at least a basic binary build.

File: src/Leanpkg/Build.lean [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  leanPath : String
+  -- things like `leanpkg.toml` and olean roots of dependencies that should also trigger rebuilds
+  moreDeps : List FilePath
+
+structure Context extends Config where
+  parents       : List Name := []
+  moreDepsMTime : IO.FS.SystemTime
+
+deriving instance Inhabited for IO.FS.SystemTime
+deriving instance Inhabited for Task
```
Comment (tydeu) [inline src/Leanpkg/Build.lean:30]: Should these instances be moved into the Lean core? Since they are not really specific to Leanpkg itself?

Comment (Kha) [reply src/Leanpkg/Build.lean]: Yes, they should

**Comment by Kha** [comment]
Tested successfully on `Init` and `mathlib4` (which has a `Makefile`, so most new code is skipped anyway), so this should be ready to go.

**Pull request merged**